src/base/perf.hpp

```diff
@@ -0,0 +1,662 @@
+//      2. static initializer for statset_definition
+//      3. automatic timer statistics conversions
+//
+//  usage:
+//
+//    // define statistics ID's; we want these const to let compiler replace them
+//    // note - stat_defition will check provided id's are successive starting with 0
+//    static const cubperf::stat_id STAT_COUNTER = 0;
+//    static const cubperf::stat_id STAT_TIMER = 1;
+//    static const cubperf::stat_id STAT_COUNTER_AND_TIMER = 2;
+//    // define your set of statistics:
+//    static const cubperf::stat_definition Mystat_definition =
+//      {
+//        cubperf::stat_definition (STAT_COUNTER, cubperf::stat_definition::COUNTER, "c_counter"),
+//        cubperf::stat_definition (STAT_TIMER, cubperf::stat_definition::TIMER, "t_timer"),
+//        cubperf::stat_definition (STAT_COUNTER_AND_TIMER, cubperf::stat_definition::COUNTER_AND_TIMER, "ct_counter"
+//                                  "ct_timer")
+//      };
+//
+//    // a statistics set of values can be created using definition
+//    cubperf::statset& my_statset = Mystat_definition.create_statset ();
+//    // an atomic set can be created with same definition
+//    cubperf::atomic_statset& my_atomic_statset = Mystat_definition.create_atomic_statset ();
+//
+//    // collecting statistics is supervised by definition
+//    Mystat_definition.increment (my_statset, STAT_COUNTER, 10);  // a counter is expected
+//
+//    // same for atomic sets
+//    Mystat_definition.time_and_increment (my_atomic_statset, STAT_COUNTER_AND_TIMER); // a counter and timer expected
+//    // note: by default counter is incremented by 1;
+//    //       by default timer is incremented by duration since last time_and_increment
+//
+//    // print statistics
+//    stat_value *values = new stat_value [Mystat_definition.get_value_count ()]; // allocate array to copy values to
+//                                                                                // just for example
+//    // get statistics; timers can be automatically converted to desired unit
+//    Mystat_definition.get_stat_values_with_converted_timers<std::chrono::milliseconds> (my_atomic_statset, values);
+//
+//    // print all values
+//    for (std::size_t index = 0; index < Mystat_definition.get_value_count (); index++)
+//      {
+//        printf ("%s = %llu\n", Mystat_definition.get_value_name (index), values[index]);
+//      }
+//
+//    // note - to extend the statistics set, you only need to:
+//    //        1. add new ID.
+//    //        2. update Mystat_definition
+//    //        3. call increment/time/time_and_increment operations on new statistic
+//
+//    delete values;
+//    delete &my_statset;
+//    delete &my_atomic_statset
+//
+
+namespace cubperf
+{
+  // stat_definition - defines one statistics entry in a set
+  class stat_definition
+  {
+    public:
+
+      // possible types
+      enum type
+      {
+	COUNTER,
+	TIMER,
+	COUNTER_AND_TIMER
```

Comment:
You may also consider another metric type, "gauge". It is not so popular but it may be useful to monitor some rates such as plan cache hit ratio, page buffer hit ratio, CPU usage rate, sector/page usage and so on.
"min" and "max" for a gauge value may also help. One gauge has 3 values: the current value, min and max. 

This extended can be separated from the PR. :)